src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -0,0 +1,1241 @@
+      if (now_ms >= scan_due_ms) {
+         break;
+      }
+
+      if (server_monitor->shared.scan_requested) {
+         server_monitor->shared.scan_requested = false;
+         scan_due_ms = start_ms + server_monitor->min_heartbeat_frequency_ms;
+      }
+
+
+      sleep_duration_ms = scan_due_ms - now_ms;
+
+      if (sleep_duration_ms > 0) {
+         MONITOR_LOG (
+            server_monitor, "sleeping for %" PRId64, sleep_duration_ms);
+         mongoc_cond_timedwait (&server_monitor->shared.cond,
+                                &server_monitor->shared.mutex,
+                                sleep_duration_ms);
+      } else {
+         break;
+      }
+   }
+   bson_mutex_unlock (&server_monitor->shared.mutex);
+}
+
+/* The server monitor thread function.
+ *
+ * Server monitor must be in state MONGOC_THREAD_OFF.
+ */
+static BSON_THREAD_FUN (_server_monitor_thread, server_monitor_void)
+{
+   mongoc_server_monitor_t *server_monitor;
+   mongoc_server_description_t *description;
+   mongoc_server_description_t *previous_description;
+
+   server_monitor = (mongoc_server_monitor_t *) server_monitor_void;
+   description =
+      mongoc_server_description_new_copy (server_monitor->description);
+   previous_description = NULL;
+
+   while (true) {
+      bool cancelled = false;
+
+      bson_mutex_lock (&server_monitor->shared.mutex);
+      if (server_monitor->shared.state != MONGOC_THREAD_RUNNING) {
+         bson_mutex_unlock (&server_monitor->shared.mutex);
+         break;
+      }
+      bson_mutex_unlock (&server_monitor->shared.mutex);
+
+      mongoc_server_description_destroy (previous_description);
+      previous_description = mongoc_server_description_new_copy (description);
+      mongoc_server_description_destroy (description);
+      description = mongoc_server_monitor_check_server (
+         server_monitor, previous_description, &cancelled);
+
+      if (cancelled) {
+         mongoc_server_monitor_wait (server_monitor);
+         continue;
+      }
+
+      _server_monitor_update_topology_description (server_monitor, description);
+
+      /* Immediately proceed to the next check if the previous response was
+       * successful and included the topologyVersion field. */
+      if (description->type != MONGOC_SERVER_UNKNOWN &&
+          !bson_empty (&description->topology_version)) {
+         MONITOR_LOG (server_monitor,
+                      "immediately proceeding due to topologyVersion");
+         continue;
+      }
+
+      /* ... or the previous response included the moreToCome flag */
+      if (server_monitor->more_to_come) {
+         MONITOR_LOG (server_monitor,
+                      "immediately proceeding due to moreToCome");
+         continue;
+      }
+
+      /* ... or the server has just transitioned to Unknown due to a network
+       * error. */
+      if (_mongoc_error_is_network (&description->error) &&
+          previous_description->type != MONGOC_SERVER_UNKNOWN) {
```

Comment:
Should we check that the current description is `UNKNOWN`, or is that a given if `_mongoc_error_is_network` returns true?